### ISSUE TITLE
SA1019 silently disappears

### ISSUE THREAD

**Issue by entombedvirus** [golangci/golangci-lint#6807]
### Welcome

- [x] Yes, I'm using a binary release within 2 latest [releases](https://github.com/golangci/golangci-lint/releases). Only such installations are supported.
- [x] Yes, I've searched similar [issues on GitHub](https://github.com/golangci/golangci-lint/issues) and didn't find any.
- [x] Yes, I've read the `typecheck` section of the [FAQ](https://golangci-lint.run/docs/welcome/faq/#why-do-you-have-typecheck-errors).
- [x] Yes, I've tried with the standalone [linter](https://golangci-lint.run/docs/linters/) if available (e.g., gocritic, go vet, etc.).
- [x] I agree to follow this project's [Code of Conduct](https://github.com/golangci/golangci-lint?tab=coc-ov-file)

### How did you install golangci-lint?

go install

### Description of the problem

When a non-initial package has cached facts for some fact analyzers but not for others, golangci-lint loses the cached facts, and every analyzer that depends on them produces false negatives. The false negatives are then saved into the issue cache of the dependent packages and repeat on every run until the source changes.

The most visible symptom is `staticcheck` SA1019: a call to a `Deprecated:` function in another package is not reported, and if the call has a `//nolint:staticcheck` directive, `nolintlint` reports it as unused. The same happens to every analyzer that imports facts from a dependency (`exhaustive`, `printf`, custom plugins). I believe this is the remaining cause behind the reports in #3228 and #5979 that survived #6683.

<details>

**Cause**

In `pkg/goanalysis/runner_loadingpackage.go`, `loadImportedPackageWithFacts` first loads the package types from export data and calls `loadCachedFacts` for every action. Each action's facts are resolved (`loadPersistedFacts`) against the objects of that export-data `types.Package`. If any action has no cached facts, the function then creates a new `types.Package`, calls `loadFromSource`, and returns. The facts already loaded by the other actions still refer to objects of the discarded package. `inheritFacts` and `exportedFrom` compare `obj.Pkg() == dep.Package.Types` by pointer, and `ImportObjectFact` looks facts up by the new objects, so the facts are never found.

The partial state is easy to reach: an interrupted run that had persisted facts for some analyzers only, an analyzer that failed on the package, a change in the enabled linter set (used in the reproduction below), or a cache directory that was partially cleaned. The reproduction shows the mechanism with a linter-set change, but in our CI the trigger is interrupted runs on a shared cache, and the false negatives then persist for days.

**Debug output confirming the mixed state** (`GL_DEBUG=goanalysis/facts/cache` on the second run below):

    [goanalysis/facts/cache] Loaded 1 cached facts for package "a" and analyzer fact_deprecated
    [goanalysis/facts/cache] No cached facts for package "a" and analyzer printf
    [goanalysis/facts/cache] Loading of facts for printf@example.com/repro/a failed, analyze it from source later

`fact_deprecated` has its fact for `a.Old`, `printf` has none, `a` is reloaded from source, and SA1019 in `b` no longer sees the deprecation.

**Fix**

I have a small patch ([golangci-lint-stale-facts-v2.13.2.patch](https://github.com/user-attachments/files/32398161/golangci-lint-stale-facts-v2.13.2.patch) about 40 lines in `runner_loadingpackage.go` and `runner_action_cache.go`) that re-resolves the already loaded cached facts against the source types after `loadFromSource`, and falls back to `needAnalyzeSource` when that fails. It fixes the reproduction and has been running on our CI. I am happy to open a PR if the approach is acceptable.

</details>

### Version of golangci-lint

golangci-lint has version 2.13.2 built with go1.26.4

### Configuration

<details>

```yml
# cfg1.yml
version: "2"
linters:
  default: none
  enable:
    - staticcheck
    - nolintlint
  settings:
    nolintlint:
      require-specific: true
```
```yml
# cfg2.yml: identical, plus govet
version: "2"
linters:
  default: none
  enable:
    - staticcheck
    - nolintlint
    - govet
  settings:
    nolintlint:
      require-specific: true
```

</details>

### Go environment

<details>

```console
go version
go version go1.26.4 darwin/arm64


AR='ar'
CC='clang'
CGO_CFLAGS='-O2 -g'
CGO_CPPFLAGS=''
CGO_CXXFLAGS='-O2 -g'
CGO_ENABLED='1'
CGO_FFLAGS='-O2 -g'
CGO_LDFLAGS='-O2 -g'
CXX='clang++'
GCCGO='gccgo'
GO111MODULE=''
GOARCH='arm64'
GOARM64='v8.0'
GOAUTH='netrc'
GOBIN=''
GOCACHE='/Users/USER/Library/Caches/go-build'
GOCACHEPROG=''
GODEBUG=''
GOENV='/Users/USER/Library/Application Support/go/env'
GOEXE=''
GOEXPERIMENT=''
GOFIPS140='off'
GOFLAGS=''
GOGCCFLAGS='-fPIC -arch arm64 -pthread -fno-caret-diagnostics -Qunused-arguments -Wl,--no-gc-sections -fmessage-length=0 -ffile-prefix-map=/var/folders/13/gk85ycz15r1cb4_y0h0tvz3c0000gn/T/go-build1838194713=/tmp/go-build -gno-record-gcc-switches -fno-common'
GOHOSTARCH='arm64'
GOHOSTOS='darwin'
GOINSECURE=''
GOMOD='/Users/USER/work/sierra/go.mod'
GOMODCACHE='/Users/USER/go/pkg/mod'
GONOPROXY='go.sierra.ai,github.com/sierra-inc/*'
GONOSUMDB='go.sierra.ai,github.com/sierra-inc/*'
GOOS='darwin'
GOPATH='/Users/USER/go'
GOROOT='/Users/USER/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.26.4.darwin-arm64'
GOSUMDB='sum.golang.org'
GOTELEMETRY='local'
GOTELEMETRYDIR='/Users/USER/Library/Application Support/go/telemetry'
GOTMPDIR=''
GOTOOLCHAIN='auto'
GOTOOLDIR='/Users/USER/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.26.4.darwin-arm64/pkg/tool/darwin_arm64'
GOVCS=''
GOVERSION='go1.26.4'
GOWORK=''
PKG_CONFIG='pkg-config'
```

</details>

### Verbose output of running

<details>

```console
$ golangci-lint cache clean
$ golangci-lint run -c cfg1.yml ./b
b/b.go:7:2: SA1019: example.com/repro/a.Old is deprecated: use New. (staticcheck)
1 issues:
* staticcheck: 1

$ golangci-lint run -v -c cfg2.yml ./b
level=info msg="[lintersdb] Active 3 linters: [govet nolintlint staticcheck]"
level=info msg="[loader] Go packages loading at mode 8767 (compiled_files|deps|exports_file|imports|types_sizes|files|name) took 10.410017ms"
level=info msg="[linters_context/goanalysis] analyzers took 2.552472ms with top 10 stages: inline: 417.66µs, printf: 379.86µs, fact_deprecated: 136.143µs, ..."
level=info msg="[runner] linters took 2.574165ms with stages: goanalysis_metalinter: 2.536691ms"
0 issues.

$ golangci-lint run -c cfg2.yml ./b      # repeats: the false negative is now in the issue cache
0 issues.

$ golangci-lint cache clean && golangci-lint run -c cfg2.yml ./b   # control: cfg2 alone is fine
b/b.go:7:2: SA1019: example.com/repro/a.Old is deprecated: use New. (staticcheck)
```

</details>

### A minimal reproducible example or link to a public repository

<details>

```go
go.mod
    module example.com/repro
    go 1.26

a/a.go
    package a

    // Deprecated: use New.
    func Old() {}

    func New() {}

b/b.go
    package b

    import "example.com/repro/a"

    func Use() {
        a.Old()
    }
```

</details>

### Validation

- [x] Yes, I've included all the information above (example, version, config, etc.).

### Supporter

- [ ] I am a [sponsor/backer](https://donate.golangci.org) of this project.
